### PR TITLE
Prepare Data How-To

### PR DESCRIPTION
**Pull request by luisquintanilla**
[WIP] Prepare Data How-To

[Internal Review Link](https://review.docs.microsoft.com/en-us/dotnet/machine-learning/how-to-guides/prepare-data-ml-net?branch=pr-en-us-11627)

[AB#1485646](https://dev.azure.com/mseng/677da0fb-b067-4f77-b89b-f32c12bb8617/_workitems/edit/1485646)

### CODE REVIEW

File: docs/machine-learning/how-to-guides/prepare-data-ml-net.md [hunk 1]
```diff
@@ -0,0 +1,331 @@
...
+var replacementEstimator = mlContext.Transforms.ReplaceMissingValues("Price", replacementMode: MissingValueReplacingEstimator.ReplacementMode.Mean);
+
+// Build transformer to perform operations
+ITransformer replacementTransformer = replacementEstimator.Fit(data);
+
+// Transform data
+IDataView transformedData = replacementTransformer.Transform(data);
+```
+
+ML.NET supports various [replacement modes](xref:Microsoft.ML.Transforms.MissingValueReplacingEstimator.ReplacementMode). The sample above uses the `Mean` replacement mode which will fill in the missing value with the average value for that column. The result of the replacement fills in the `Price` property for the last element in our data with 200000 since it's the average of 100000 and 300000.
```
Comment (JRAlexander) [inline docs/machine-learning/how-to-guides/prepare-data-ml-net.md]: "the average value for that column" -> "that column's average value"
"The result of the replacement" -> "The replacement's result"

File: docs/machine-learning/how-to-guides/prepare-data-ml-net.md [hunk 2]
```diff
@@ -0,0 +1,331 @@
...
+
+// Transform Data
+IDataView transformedData = binningTransformer.Transform(data);
+```
+
+The result of binning creates bin bounds of `[0,200000,Infinity]`. Therefore the resulting bins are `[0,1,1]` because the first observation is between 0-200000 and the others are greater than 200000 but less than infinity.
+
+## Work with categorical data
+
+Categorical data that is not a number needs to be converted to a number before being used to build a machine learning model.
```
Comment (JRAlexander) [inline docs/machine-learning/how-to-guides/prepare-data-ml-net.md]: "Categorical data that is not a number" -> "Non-numeric categorical data"

File: docs/machine-learning/how-to-guides/prepare-data-ml-net.md [hunk 3]
```diff
@@ -0,0 +1,331 @@
...
+    },
+    new HomeData
+    {
+        NumberOfBedrooms=6f,
+        Price=600000f
+    }
+};
+```
+
+Normalize the data into bins using the [`NormalizeBinning`](xref:Microsoft.ML.NormalizationCatalog.NormalizeBinning*) method. The `maximumBinCount` parameter enables you to specify how many bins you want to put your data into. In this example, data will be put into two bins.
```
Comment (JRAlexander) [inline docs/machine-learning/how-to-guides/prepare-data-ml-net.md]: "specify how many bins you want to put your data into" -> "specify the number of bins needed to classify your data."

File: docs/machine-learning/how-to-guides/prepare-data-ml-net.md [hunk 4]
```diff
@@ -0,0 +1,331 @@
...
+IDataView transformedData = textTransformer.Transform(data);
+```
+
+The resulting transform would convert the text values in the `Description` column to a numerical vector that looks similar to the output below:
+
+```text
+[ 0.2041241, 0.2041241, 0.2041241, 0.4082483, 0.2041241, 0.2041241, 0.2041241, 0.2041241, 0.2041241, 0.2041241, 0.2041241, 0.2041241, 0.2041241, 0.2041241, 0.2041241, 0.2041241, 0.2041241, 0.2041241, 0.2041241, 0.2041241, 0.2041241, 0, 0, 0, 0, 0.4472136, 0.4472136, 0.4472136, 0.4472136, 0.4472136, 0 ]
+```
+
+More complex text processing steps can be combined to remove noise and potentially reduce the amount of resources required for processing.
```
Comment (JRAlexander) [inline docs/machine-learning/how-to-guides/prepare-data-ml-net.md]: Combine complex text processing steps to remove noise and potentially reduce the amount of required processing resources.

File: docs/machine-learning/how-to-guides/prepare-data-ml-net.md [hunk 5]
```diff
@@ -0,0 +1,331 @@
...
+
+```csharp
+// Define text transform estimator
+var textEstimator = mlContext.Transforms.Text.NormalizeText("Description")
+    .Append(mlContext.Transforms.Text.TokenizeIntoWords("Description"))
+    .Append(mlContext.Transforms.Text.RemoveDefaultStopWords("Description"))
+    .Append(mlContext.Transforms.Text.FeaturizeText("Description"));
+```
+
+Using the first entry as an example, below is a detailed description of the results produced by the transformation steps defined in the code sample above:
```
Comment (JRAlexander) [inline docs/machine-learning/how-to-guides/prepare-data-ml-net.md]: "below" -> "the following"
"defined in the code sample above" -> "defined in the previous code sample"

File: docs/machine-learning/how-to-guides/prepare-data-ml-net.md [hunk 6]
```diff
@@ -0,0 +1,331 @@
+---
+title: "How-To: Prepare Data with ML.NET"
+description: Learn how to use transforms in ML.NET to manipulate and prepare data for additional processing or model building.
+ms.date: 04/05/2019
+ms.custom: mvc,how-to
```
Comment (JRAlexander) [inline docs/machine-learning/how-to-guides/prepare-data-ml-net.md]: add:
author: luisquintanilla
ms.author: luquinta

**Review by JRAlexander — CHANGES_REQUESTED** [review]
Looks good. A few small changes then you'll be good to go.

File: docs/machine-learning/how-to-guides/prepare-data-ml-net.md [hunk 7]
```diff
@@ -0,0 +1,309 @@
...
+author: luisquintanilla
+ms.author: luquinta
+ms.date: 04/09/2019
+ms.custom: mvc,how-to
+#Customer intent: As a developer, I want to know how I can transform and prepare data with ML.NET using transforms
+---
+
+# How-To: Prepare Data with ML.NET
+
+This how-to shows how to use transformers in ML.NET to prepare data for additional processing or building a model.
```
Comment (natke) [inline docs/machine-learning/how-to-guides/prepare-data-ml-net.md]: I think this would be a good place to describe features in general, and the requirements ML.NET places on the Features column.

File: docs/machine-learning/how-to-guides/prepare-data-ml-net.md [hunk 8]
```diff
@@ -0,0 +1,309 @@
...
+ITransformer replacementTransformer = replacementEstimator.Fit(data);
+
+// Transform data
+IDataView transformedData = replacementTransformer.Transform(data);
+```
+
+ML.NET supports various [replacement modes](xref:Microsoft.ML.Transforms.MissingValueReplacingEstimator.ReplacementMode). The sample above uses the `Mean` replacement mode which will fill in the missing value with that column's average value. The replacement
+'s result fills in the `Price` property for the last element in our data with 200,000 since it's the average of 100,000 and 300,000. 
+
+## Filter data
```
Comment (natke) [inline docs/machine-learning/how-to-guides/prepare-data-ml-net.md]: We should distinguish between operations that are part of a model and those that are separate i.e. those that are estimators/transforms and those that operate on the data directly, like these ones.

**Review by natke — CHANGES_REQUESTED** [review]

File: docs/machine-learning/how-to-guides/prepare-data-ml-net.md [hunk 9]
```diff
@@ -0,0 +1,290 @@
...
+
+# How-To: Prepare Data with ML.NET
+
+Learn how to use transformers in ML.NET to prepare data for additional processing or building a model.
+
+Data can be messy. In addition, ML.NET machine learning algorithms require input to be a numerical vector. ML.NET provides various transforms to help clean your data and get it into the appropriate format. 
+
+## Replace missing values
+
+Missing values are a common occurrence in datasets. One approach to dealing with missing values is to replace them with a placeholder or another meaningful value such as the mean value in the data.
```
Comment (sfilipi) [inline docs/machine-learning/how-to-guides/prepare-data-ml-net.md]: >placeholder [](start = 122, length = 11)

I'd say default value

File: docs/machine-learning/how-to-guides/prepare-data-ml-net.md [hunk 10]
```diff
@@ -0,0 +1,290 @@
...
+Notice that the last element in our list has a missing value for `Price`. To replace the missing values in the `Price` column, use the [`ReplaceMissingValues`](xref:Microsoft.ML.ExtensionsCatalog.ReplaceMissingValues*) method to fill in that missing value.
+
+> [!IMPORTANT]
+> [`ReplaceMissingValue`](xref:Microsoft.ML.ExtensionsCatalog.ReplaceMissingValues*) only works with numerical data.
+
+```csharp
+// Define replacement estimator
+var replacementEstimator = mlContext.Transforms.ReplaceMissingValues("Price", replacementMode: MissingValueReplacingEstimator.ReplacementMode.Mean);
+
+// Build transformer to perform operations
```
Comment (sfilipi) [inline docs/machine-learning/how-to-guides/prepare-data-ml-net.md]: >Build transformer to perform operations [](start = 3, length = 39)

Fitting the data to this estimator.

File: docs/machine-learning/how-to-guides/prepare-data-ml-net.md [hunk 11]
```diff
@@ -0,0 +1,290 @@
...
+    },
+    new HomeData
+    {
+        NumberOfBedrooms = 1f,
+        Price = 100000f
+    }
+};
+```
+
+Normalize the data using min-max normalization using the [`NormalizeMinMax`](xref:Microsoft.ML.NormalizationCatalog.NormalizeMinMax*) method.
```
Comment (sfilipi) [inline docs/machine-learning/how-to-guides/prepare-data-ml-net.md]: >Normalize [](start = 0, length = 9)

Maybe we should say why normalizer the data. Typically normalization is done so the algorithm converges quicker. 
This is typically needed for linear trainers. 

[https://docs.microsoft.com/en-us/azure/machine-learning/studio-module-reference/normalize-data](https://docs.microsoft.com/en-us/azure/machine-learning/studio-module-reference/normalize-data)

File: docs/machine-learning/how-to-guides/prepare-data-ml-net.md [hunk 12]
```diff
@@ -0,0 +1,290 @@
...
+var minMaxEstimator = mlContext.Transforms.NormalizeMinMax("Price");
+
+// Build ITransformer that performs operation
+ITransformer minMaxTransformer = minMaxEstimator.Fit(data);
+
+// Transform data
+IDataView transformedData = minMaxTransformer.Transform(data);
+```
+
+The original price values `[200000,100000]` are converted to `[ 1, 0.5 ]` using the `MinMax` normalization formula.
```
Comment (sfilipi) [inline docs/machine-learning/how-to-guides/prepare-data-ml-net.md]: >`[ 1, 0.5 ]` using the `MinMax` normalization formula [](start = 61, length = 53)

might be helpful to say that min-max rescales in the [0, 1] interval, so it makes sense why they get mapped to those values.

File: docs/machine-learning/how-to-guides/prepare-data-ml-net.md [hunk 13]
```diff
@@ -0,0 +1,290 @@
...
+
+### Binning
+
+[Binning](https://en.wikipedia.org/wiki/Data_binning) converts continuous values into a discrete representation. For example, suppose one of your features is age. Instead of using the actual age value, with binning you can create ranges for that value. 0-18 could be one bin, another could be 19-35 and so on.
```
Comment (sfilipi) [inline docs/machine-learning/how-to-guides/prepare-data-ml-net.md]: >create ranges for that value. 0-18 could be one bin, another could be 19-35 and so on. [](start = 223, length = 86)

This might give them the illusion that they can control the bin upperbounds, which they can't... they can just control the number of bins..

File: docs/machine-learning/how-to-guides/prepare-data-ml-net.md [hunk 14]
```diff
@@ -0,0 +1,290 @@
...
+var textEstimator  = mlContext.Transforms.Text.FeaturizeText("Description");
+
+// Build ITransformer that performs operation
+ITransformer textTransformer = textEstimator.Fit(data);
+
+// Transform data
+IDataView transformedData = textTransformer.Transform(data);
+```
+
+The resulting transform would convert the text values in the `Description` column to a numerical vector that looks similar to the output below:
```
Comment (sfilipi) [inline docs/machine-learning/how-to-guides/prepare-data-ml-net.md]: >numerical vector [](start = 87, length = 16)

if you wanted to add a bit more info, you could put:

a numerical vector representing the normalized counts of the n-grams (groups of characters) appearing in the text.

File: docs/machine-learning/how-to-guides/prepare-data-ml-net.md [hunk 15]
```diff
@@ -0,0 +1,290 @@
...
+```
+
+Combine complex text processing steps to remove noise and potentially reduce the amount of required processing resources as needed.
+
+```csharp
+// Define text transform estimator
+var textEstimator = mlContext.Transforms.Text.NormalizeText("Description")
+    .Append(mlContext.Transforms.Text.TokenizeIntoWords("Description"))
+    .Append(mlContext.Transforms.Text.RemoveDefaultStopWords("Description"))
+    .Append(mlContext.Transforms.Text.FeaturizeText("Description"));
```
Comment (sfilipi) [inline docs/machine-learning/how-to-guides/prepare-data-ml-net.md]: I would not put those together, because the FeaturizeText does the tokenization, removing of stop words etc. 

Are you trying to give a pipeline that is equivalent to FeaturizeText? That would be a helpful thing to do, when the users are debugging their pipelines, and don;t want blackboxes.


Then you can put the following:
Text.NormalizeText
Text.TokenizeIntoWords
Text.RemoveDefaultStopWords
Text.ProduceWordBags
Text.ProduceNgrams

you'll need to concatenate the output of the last two. 

Transforms.NormalizeLpNorm

Basically do what TextFeaturizingEstimator.Fit is doing

Comment (sfilipi) [reply docs/machine-learning/how-to-guides/prepare-data-ml-net.md]: The one in the LDA sample is a very similar pipeline:
[https://github.com/dotnet/machinelearning/blob/master/docs/samples/Microsoft.ML.Samples/Dynamic/Transforms/Text/LatentDirichletAllocation.cs#L32](https://github.com/dotnet/machinelearning/blob/master/docs/samples/Microsoft.ML.Samples/Dynamic/Transforms/Text/LatentDirichletAllocation.cs#L32) 

  var pipeline = mlContext.Transforms.Text.NormalizeText("NormalizedText", "Text")
                .Append(mlContext.Transforms.Text.TokenizeIntoWords("Tokens", "NormalizedText"))
                .Append(mlContext.Transforms.Text.RemoveDefaultStopWords("Tokens"))
                .Append(mlContext.Transforms.Conversion.MapValueToKey("Tokens"))
                .Append(mlContext.Transforms.Text.ProduceNgrams("Tokens"))
                .Append(mlContext.Transforms.Text.LatentDirichletAllocation("Features", "Tokens", numberOfTopics: 3)); 

---
In reply to: [277379997](https://github.com/dotnet/docs/pull/11627#discussion_r277379997) [](ancestors = 277379997)

**Review by sfilipi — CHANGES_REQUESTED** [review]
:clock1:

**Review by natke — APPROVED** [review]

File: docs/machine-learning/how-to-guides/prepare-data-ml-net.md [hunk 16]
```diff
@@ -0,0 +1,302 @@
...
+    },
+    new HomeData
+    {
+        NumberOfBedrooms=6f,
+        Price=600000f
+    }
+};
+```
+
+Normalize the data into bins using the [`NormalizeBinning`](xref:Microsoft.ML.NormalizationCatalog.NormalizeBinning*) method. The `maximumBinCount` parameter enables you to specify the numer of bins needed to classify your data. In this example, data will be put into two bins.
```
Comment (JRAlexander) [inline docs/machine-learning/how-to-guides/prepare-data-ml-net.md]:
```suggestion
Normalize the data into bins using the [`NormalizeBinning`](xref:Microsoft.ML.NormalizationCatalog.NormalizeBinning*) method. The `maximumBinCount` parameter enables you to specify the number of bins needed to classify your data. In this example, data will be put into two bins.  
```

**Review by JRAlexander — APPROVED** [review]
